Mathlib/AlgebraicGeometry/AffineScheme.lean

```diff
@@ -747,6 +775,46 @@ theorem self_le_basicOpen_union_iff (s : Set Γ(X, U)) :
 
 end IsAffineOpen
 
+open _root_.PrimeSpectrum in
+/-- The restriction of `Spec.map f` to a basic open `D(r)` is isomorphic to `Spec.map` of the
+localization of `f` away from `r`. -/
+noncomputable
+def SpecMapRestrictBasicOpenIso {R S : CommRingCat} (f : R ⟶ S) (r : R) :
+    Arrow.mk (Spec.map f ∣_ (PrimeSpectrum.basicOpen r)) ≅
+      Arrow.mk (Spec.map <| CommRingCat.ofHom (Localization.awayMap f r)) := by
+  letI e₁ : Localization.Away r ≃ₐ[R] Γ(Spec R, basicOpen r) :=
+    IsLocalization.algEquiv (Submonoid.powers r) _ _
+  letI e₂ : Localization.Away (f r) ≃ₐ[S] Γ(Spec S, basicOpen (f r)) :=
+    IsLocalization.algEquiv (Submonoid.powers (f r)) _ _
+  refine Arrow.isoMk ?_ ?_ ?_
+  · exact (Spec (.of S)).isoOfEq (comap_basicOpen _ _) ≪≫
+      (IsAffineOpen.Spec_basicOpen (f r)).isoSpec ≪≫ Scheme.Spec.mapIso e₂.toCommRingCatIso.op
+  · exact (IsAffineOpen.Spec_basicOpen r).isoSpec ≪≫ Scheme.Spec.mapIso e₁.toCommRingCatIso.op
+  · have := AlgebraicGeometry.IsOpenImmersion.of_isLocalization
+      (S := (Localization.Away r)) r
+    rw [← cancel_mono (Spec.map (CommRingCat.ofHom (algebraMap R (Localization.Away r))))]
+    simp only [Arrow.mk_left, Arrow.mk_right, Functor.id_obj, Scheme.isoOfEq_rfl, Iso.refl_trans,
+      Iso.trans_hom, Functor.mapIso_hom, Iso.op_hom, Iso.symm_hom, RingEquiv.toCommRingCatIso_inv,
+      Scheme.Spec_map, Quiver.Hom.unop_op, Arrow.mk_hom, Category.assoc,
+      ← Spec.map_comp]
+    show _ ≫ Spec.map ((e₂.toRingHom.comp (Localization.awayMap f r)).comp (algebraMap R _))
+      = _ ≫ _ ≫ Spec.map (e₁.toRingHom.comp (algebraMap R _))
+    rw [RingHom.comp_assoc]
+    conv => enter [1,2,1,2]; tactic => exact IsLocalization.map_comp _
+    rw [← RingHom.comp_assoc]
+    conv => enter [1,2,1,1]; tactic => exact e₂.toAlgHom.comp_algebraMap
+    conv => enter [2,2,2,1]; tactic => exact e₁.toAlgHom.comp_algebraMap
+    show _ ≫ Spec.map (f ≫ (Scheme.ΓSpecIso S).inv ≫
+      (Spec S).presheaf.map (homOfLE le_top).op) =
+      Spec.map f ∣_ basicOpen r ≫ _ ≫ Spec.map ((Scheme.ΓSpecIso R).inv ≫
+      (Spec R).presheaf.map (homOfLE le_top).op)
+    simp only [IsAffineOpen.isoSpec_hom, homOfLE_leOfHom, Spec.map_comp, Category.assoc,
+      Scheme.Opens.toSpecΓ_SpecMap_map_assoc, Scheme.Opens.toSpecΓ_top, Scheme.homOfLE_ι_assoc,
+      morphismRestrict_ι_assoc]
+    simp only [← SpecMap_ΓSpecIso_hom, ← Spec.map_comp, Category.assoc, Iso.inv_hom_id,
+      Category.comp_id, Category.id_comp]
+    rfl
```

Comment:
I wonder if one day `egg` with an AG tailored rule set can solve these goals.

Reply:
I think the main friction comes from (again!) the switch from category theoretic arrows to RingHoms and back.